pkg/reconciler/cache/replication/replication_reconcile.go

```diff
@@ -19,102 +19,130 @@ package replication
 import (
 	"context"
 	"fmt"
+	"reflect"
 	"strings"
 
 	kcpcache "github.com/kcp-dev/apimachinery/pkg/cache"
 	"github.com/kcp-dev/logicalcluster/v2"
 
 	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
 
 	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
 )
 
-func (c *controller) reconcile(ctx context.Context, grKey string) error {
-	keyParts := strings.Split(grKey, "::")
+func (c *controller) reconcile(ctx context.Context, gvrKey string) error {
+	keyParts := strings.Split(gvrKey, "::")
 	if len(keyParts) != 2 {
-		return fmt.Errorf("incorrect key: %v, expected group.resource::key", grKey)
+		return fmt.Errorf("incorrect key: %v, expected group.version.resource::key", gvrKey)
 	}
 	switch keyParts[0] {
 	case apisv1alpha1.SchemeGroupVersion.WithResource("apiexports").String():
-		return c.reconcileAPIExports(ctx, keyParts[1], apisv1alpha1.SchemeGroupVersion.WithResource("apiexports"))
+		return c.reconcileObject(ctx,
+			keyParts[1],
+			apisv1alpha1.SchemeGroupVersion.WithResource("apiexports"),
+			apisv1alpha1.SchemeGroupVersion.WithKind("APIExport"),
+			func(gvr schema.GroupVersionResource, cluster, namespace, name string) (interface{}, error) {
```

Comment:
Can we use generics to keep this strongly typed?

Reply:
mhm, I think it would require refactoring because you cannot define a type parameter on a method (in that case that would be reconcileObject). I'm not sure if this is worth it. What would that give us? This code doesn't need to know about the type because it translates an object to an unstructured object anyway.

Reply:
I haven't used generics in go yet, I experimented with https://go.dev/play/p/nI3bsSm_MRZ